subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c

```diff
@@ -0,0 +1,242 @@
+extern struct wpa_global *global;
+
+
+LOG_MODULE_REGISTER(wifi_mgmt_ext, CONFIG_WIFI_MGMT_EXT_LOG_LEVEL);
+
+/* helper functions {*/
+
+static inline struct wpa_supplicant *get_wpa_s_handle(const struct device *dev)
+{
+	struct wpa_supplicant *wpa_s = NULL;
+	int ret = k_sem_take(&wpa_supplicant_ready_sem, K_SECONDS(2));
+
+	if (ret) {
+		LOG_ERR("%s: WPA supplicant not ready: %d", __func__, ret);
+		return NULL;
+	}
+
+	k_sem_give(&wpa_supplicant_ready_sem);
+
+	wpa_s = wpa_supplicant_get_iface(global, dev->name);
+	if (!wpa_s) {
+		LOG_ERR("%s: Unable to get wpa_s handle for %s\n", __func__, dev->name);
+		return NULL;
+	}
+
+	return wpa_s;
+}
+
+static inline void band_to_freq(struct wifi_credentials_personal *creds, struct wpa_ssid *ssid)
+{
+	/** 2.4 GHz channels: 1-14 */
+	const int channels_2_4ghz[] = {
+		2412, 2417, 2422, 2427,
+		2432, 2437, 2442, 2447,
+		2452, 2457, 2462, 2467,
+		2472, 2484
+	};
+
+	/** 5 GHz channels:
+	 *   36,  40,  44,  48,
+	 *   52,  56,  60,  62,
+	 *   64, 100, 104, 108,
+	 *  112, 116, 120, 124,
+	 *  128, 132, 136, 140,
+	 *  149, 153, 157, 161,
+	 *  165
+	 */
+	const int channels_5ghz[] = {
+		5180, 5200, 5220, 5240,
+		5260, 5280, 5300, 5310,
+		5320, 5500, 5520, 5540,
+		5560, 5580, 5600, 5620,
+		5640, 5660, 5680, 5700,
+		5745, 5765, 5785, 5805,
+		5825
+	};
+
+	if (!(creds->header.flags & (WIFI_CREDENTIALS_FLAG_2_4GHz | WIFI_CREDENTIALS_FLAG_5GHz))) {
+		/* there are no flags indicating band - skip */
+		return;
+	}
+	if ((creds->header.flags & (WIFI_CREDENTIALS_FLAG_2_4GHz | WIFI_CREDENTIALS_FLAG_5GHz)) ==
+	    (WIFI_CREDENTIALS_FLAG_2_4GHz | WIFI_CREDENTIALS_FLAG_5GHz)) {
+		/* both bands are indicated - skip */
+		return;
+	}
+
+	if (creds->header.flags & WIFI_CREDENTIALS_FLAG_2_4GHz) {
+		ssid->freq_list = os_zalloc((ARRAY_SIZE(channels_2_4ghz) + 1) * sizeof(int));
+		if (!ssid->freq_list) {
+			LOG_ERR("unable to allocate buffer for freq_list");
+			return;
+		}
+		memcpy(ssid->freq_list, channels_2_4ghz, ARRAY_SIZE(channels_2_4ghz) * sizeof(int));
+	}
+
+	if (creds->header.flags & WIFI_CREDENTIALS_FLAG_5GHz) {
+		ssid->freq_list = os_zalloc((ARRAY_SIZE(channels_5ghz) + 1) * sizeof(int));
+		if (!ssid->freq_list) {
+			LOG_ERR("unable to allocate buffer for freq_list");
+			return;
+		}
+		memcpy(ssid->freq_list, channels_5ghz, ARRAY_SIZE(channels_5ghz) * sizeof(int));
+	}
+}
+
+static inline int copy_psk(enum wifi_security_type security,
+			   const uint8_t *password,
+			   size_t password_len,
+			   struct wpa_ssid *ssid)
+{
+	if (!password) {
+		return 0;
+	}
+	switch (security) {
+	case WIFI_SECURITY_TYPE_NONE:
+		ssid->key_mgmt = WPA_KEY_MGMT_NONE;
+		break;
+	case WIFI_SECURITY_TYPE_SAE:
+		ssid->key_mgmt = WPA_KEY_MGMT_SAE;
+		str_clear_free(ssid->sae_password);
+		ssid->sae_password = dup_binstr(password, password_len);
+
+		if (ssid->sae_password == NULL) {
+			LOG_ERR("%s:Failed to copy sae_password\n", __func__);
+			return -ENOMEM;
+		}
+		break;
+	case WIFI_SECURITY_TYPE_PSK:
```

Comment:
If the fallthrough is intentional, please add a comment saying it is

Reply:
done